plutarch-ledger-api/src/Plutarch/LedgerApi/Interval.hs

```diff
@@ -273,10 +296,112 @@ pmember ::
+  Term s (PInterval a :--> PBool)
+pisEmpty = phoistAcyclic $ plam $ \i -> unTermCont $ do
+  unpacked <- tcont $ pletFields @'["from", "to"] i
+  let lowerBound = getField @"from" unpacked
+  let upperBound = getField @"to" unpacked
+  let inclusiveLowerBound = pinclusiveLowerBound # lowerBound
+  let inclusiveUpperBound = pinclusiveUpperBound # upperBound
+  pure $ inclusiveLowerBound #> inclusiveUpperBound
+
+{- | Turn a 'PLowerBound' into a single inclusive bounding value.
+
+@since WIP
+-}
+pinclusiveLowerBound ::
+  forall (a :: S -> Type) (s :: S).
+  (PIsData a, PCountable a) =>
+  Term s (PLowerBound a :--> PExtended a)
+pinclusiveLowerBound = phoistAcyclic $ plam $ \lb -> unTermCont $ do
+  unpacked <- tcont $ pletFields @'["_0", "_1"] lb
+  let extended = getField @"_0" unpacked
+  let closure = getField @"_1" unpacked
+  pure $
+    pif
+      closure
+      -- We are already closed
+      extended
+      ( pmatch extended $ \case
+          -- Open at a finite value, get its successor
+          PFinite t -> pletFields @'["_0"] t $ \unpackedT ->
+            pcon . PFinite $ pdcons # pdata (psuccessor # getField @"_0" unpackedT) # pdnil
+          -- We have an infinity, who cares
+          _ -> extended
+      )
+
+{- | Turn a 'PUpperBound' into a single inclusive bounding value.
+
+@since WIP
+-}
+pinclusiveUpperBound ::
+  forall (a :: S -> Type) (s :: S).
+  (PIsData a, PEnumerable a) =>
+  Term s (PUpperBound a :--> PExtended a)
+pinclusiveUpperBound = phoistAcyclic $ plam $ \ub -> unTermCont $ do
+  unpacked <- tcont $ pletFields @'["_0", "_1"] ub
+  let extended = getField @"_0" unpacked
+  let closure = getField @"_1" unpacked
+  pure $
+    pif
+      closure
+      -- We are already closed
+      extended
+      ( pmatch extended $ \case
+          -- Open at a finite value, get its predecessor
+          PFinite t -> pletFields @'["_0"] t $ \unpackedT ->
+            pcon . PFinite $ pdcons # pdata (ppredecessor # getField @"_0" unpackedT) # pdnil
+          -- We have an infinity, who cares
+          _ -> extended
+      )
+
+{- | @'pcontains' # i1 # i2@ is true if @i2@ is entirely contained in @i1@: more
+specifically, if for any @s@, if @'pmember' # s # i2@, then @'pmember' # s #
+i1@.
+
+@since WIP
+-}
+pcontains ::
+  forall (a :: S -> Type) (s :: S).
+  (PIsData a, PEnumerable a) =>
+  Term
+    s
+    ( PInterval a
+        :--> PInterval a
+        :--> PBool
+    )
+pcontains = phoistAcyclic $ plam $ \i1 i2 ->
+  pif
+    (pisEmpty # i2)
+    (pcon PTrue) -- the empty interval is in everything
+    ( pif
+        (pisEmpty # i1)
+        (pcon PFalse) -- the empty interval contains nothing
+        (go i1 i2)
+    )
+  where
+    go ::
+      forall (s' :: S).
+      Term s' (PInterval a) ->
+      Term s' (PInterval a) ->
+      Term s' PBool
+    go i1 i2 = unTermCont $ do
+      unpackedI1 <- tcont $ pletFields @'["from", "to"] i1
+      unpackedI2 <- tcont $ pletFields @'["from", "to"] i2
```

Comment:
This unpacking computation is duplicated. `pisEmpty` unpacks values and this part does it again from scratch, unpacking same thing twice.